### PR TITLE
Dydx fast withdrawal api 

### PR DESCRIPTION
**Pull request by Akhilleshgoswami**

### CODE REVIEW

File: components/order_manager.py [hunk 1]
```diff
@@ -1,26 +1,46 @@
...
-            {"status": "CANCEL"}
-        )
+    def update_order_data(self, id):
+        self.update_on_firebase(id, "dydx_orders", "CANCEL")
 
+    def store_withdrawal(self, withdrawal_data):
+        self.store_on_firebase("withdrawal", withdrawal_data)
+
+    def update_withdrawal_data(self, id):
+        self.update_on_firebase(id, "withdrawal", "CONFIRMED")
```
Comment (Prithviraj8) [inline components/order_manager.py]: "dydx_orders", "CANCEL"
"withdrawal", "CONFIRMED": pass as params

File: dydx_withdrawal/views.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            result["message"]["address"] = data["user_address"]
+            db_ref.store_withdrawal(result["message"])
+            return Response(result, status.HTTP_200_OK)
+        except DydxApiError or ValueError as e:
+            e = vars(e)
+            result["error"] = e["msg"]["errors"][0]["msg"]
+            error_codes = ErrorCodes
+            if error_codes.signature_error.value == result["error"]:
+                return Response(result, status.HTTP_400_BAD_REQUEST)
+            return Response(result, status.HTTP_400_BAD_REQUEST)
```
Comment (Prithviraj8) [inline dydx_withdrawal/views.py:39]: HTTP_400_BAD_REQUEST should only be returned when the request body has some incorrect data.

Comment (Prithviraj8) [reply dydx_withdrawal/views.py:39]: return Internal server error

File: dydx_withdrawal/views.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+            withdrawal_data = withdrawal_ref.fast_withdrawal(withdrawal_data)
+            result["message"] = withdrawal_data.get('withdrawal')
+            return Response(result, status.HTTP_200_OK)
+        except DydxApiError or ValueError as e:
+            e = vars(e)
+            result["error"] = e["msg"]["errors"][0]["msg"]
+            error_codes = ErrorCodes
+            if error_codes.signature_error.value == result["error"]:
+                return Response(result, status.HTTP_400_BAD_REQUEST)
+            return Response(result, status.HTTP_400_BAD_REQUEST)
```
Comment (Prithviraj8) [inline dydx_withdrawal/views.py:59]: same

File: dydx_withdrawal/views.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+            transfer_data = transfer_ref.transfer(data)
+            result["message"] = transfer_data["data"]["transfers"]
+            return Response(result, status.HTTP_200_OK)
+        except DydxApiError or ValueError as e:
+            e = vars(e)
+            result["error"] = e["msg"]["errors"][0]["msg"]
+            error_codes = ErrorCodes
+            if error_codes.signature_error.value == result["error"]:
+                return Response(result, status.HTTP_400_BAD_REQUEST)
+            return Response(result, status.HTTP_400_BAD_REQUEST)
```
Comment (Prithviraj8) [inline dydx_withdrawal/views.py:81]: same

File: order/views.py [hunk 5]
```diff
@@ -103,6 +110,8 @@ def orders(self, request):
 
         try:
             orders = order_manager.fetch_orders(order_id=validated_data.get("order_id"))
+            if orders is None:
+                raise RuntimeError("Order id not found")
```
Comment (Prithviraj8) [inline order/views.py]: don't raise here.

File: components/order_manager.py (hunk 1 above)
Comment (Akhilleshgoswami) [reply components/order_manager.py]: bro vo kr diya he celery wali me

File: services/celery/celery.py [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+@app.task(name="check_withdrawal", default_retry_delay=4 * 60)
+def check_withdrawal():
+    dydx_withdrawal = DydxWithdrawal()
+    order_manager = OrderManager()
+    result = dydx_withdrawal.transfer({"limit": 100, "transfer_type": "WITHDRAWAL"})
+    transfers = result["data"]["transfers"]
+    for transfer in transfers:
+        if transfer["status"] == "CONFIRMED":
+            order_manager.update_withdrawal_data(transfer["id"])
```
Comment (Prithviraj8) [inline services/celery/celery.py]: we're looping over each transfer and if status is CONFIRMED we then update the status for that transfer on our firestore db ?

File: services/contracts/starkex_contract/starkex_contract.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+        self.w3 = Web3(Web3.HTTPProvider(config("WEB_PROVIDER")))
+        self.contract = self.w3.eth.contract(
+            address=config("STARKEX_ADDRESS"), abi=dydxabi
+        )
+
+    """function withdraw will withdraw funds form dydx contract"""
+
+    def withdraw(self):
+        transaction = self.contract.functions.withdraw(
+            starkKey=int(config("STRAK_PUBLIC_KEY"), 16), assetType=config("ASSET_TYPE")
```
Comment (Prithviraj8) [inline services/contracts/starkex_contract/starkex_contract.py]: please do spell checks: STRAK_PUBLIC_KEY 😅

File: services/dydx_client/admin.py [hunk 8]
```diff
@@ -30,7 +30,7 @@ def create_user(self):
     # function get_register_user is responsible register dydx_user on dydx.
     def register_user(self):
         signature = self.client.private.get_registration()
-        return signature
+        return vars(signature)
```
Comment (Prithviraj8) [inline services/dydx_client/admin.py]: safety check for None is needed here. vars(None) will break

File: order/views.py (hunk 5 above)
Comment (Akhilleshgoswami) [reply order/views.py]: have change it

File: services/celery/celery.py (hunk 6 above)
Comment (Akhilleshgoswami) [reply services/celery/celery.py]: yes

File: services/contracts/starkex_contract/starkex_contract.py (hunk 7 above)
Comment (Akhilleshgoswami) [reply services/contracts/starkex_contract/starkex_contract.py]: hnn i have change it

File: dydx_withdrawal/views.py (hunk 2 above)
Comment (Akhilleshgoswami) [reply dydx_withdrawal/views.py:39]: but if signature is wrong it's mean there is some  wrong in request body

Comment (Prithviraj8) [reply dydx_withdrawal/views.py:39]: no that doesn't constitute as 400 bad req.. 400 bad req is if you send a wrong key in the requst body.. or you miss out on sending a needed key..
wrong signature is still and internal error

Comment (Akhilleshgoswami) [reply dydx_withdrawal/views.py:39]: But this error will come if you pass wrong position if

Comment (Akhilleshgoswami) [reply dydx_withdrawal/views.py:39]: And position id is passed in req body

**Pull request merged**